### ISSUE TITLE
Support for working with raw audio

### ISSUE THREAD

**Issue by pzelasko** [pzelasko/lhotse#44]
Currently, Lhotse is somewhat based on the assumption that a user will mostly work with features such as log-mel energies. It'll be useful to have PyTorch Datasets providing raw audio as well. I'm thinking we should make it a convention to put `Waveform` in the class name of datasets that provide raw audio (e.g. `UnsupervisedWaveformDataset`). I'll try to list the steps that are needed to get there. 

- [ ] Create `Cut`s directly from `RecordingSet` (with an empty `Features` field - currently, you have to extract _some_ features to make cuts, just to ignore them later)
- [x] load_audio() in `Cut`
- [ ] load_audio() in `MixedCut`
- [x] load_audio() in `PaddingCut`
- [ ] Implement `UnsupervisedWaveformDataset`
- [ ] Implement other waveform datasets

**Comment by jimbozhang** [comment]
If nobody is working on this issue, I'll implement this.

**Comment by danpovey** [comment]
Great as far as I'm concerned!

**Comment by pzelasko** [comment]
Awesome! 

One suggestion for a starting point: it's probably the simplest if we add a `recording: Optional[Recording] = None` field to the `Cut` class, and make `features` optional as well. If a user wants a cut with audio loading capability, let them make sure they created it this way. This will avoid the manifests growing even bigger for users who do not need this capability.

**Comment by jimbozhang** [comment]
> One suggestion for a starting point: it's probably the simplest if we add a `recording: Optional[Recording] = None` field to the `Cut` class, and make `features` optional as well. If a user wants a cut with audio loading capability, let them make sure they created it this way. This will avoid the manifests growing even bigger for users who do not need this capability.

I wonder is it appropriate to take the `Recording` as a special feature:

`GeneralFeatures = Union['Features', 'Recording']`, and then define non-optional `features: GeneralFeatures` in the class `Cut`, instead of `features: Features = None; recording: Recording = None`

The motivation is that a `Cut` is defined as an example for training, which should contains one type of features and the corresponding label. The features here should be allowed in any form, not limited in spectrum. So it is a little redundancy to contain both raw audio and cepstral features in `Cut`.

**Comment by danpovey** [comment]
There are situations where you might want to access both the features and the raw recording, for different reasons.  I think it makes sense to keep the concepts separate, especially since they have different structure.

However something that does concern me, which I don't know the right way to handle, is that there might be situations where the cut doesn't have a specific channel associated with it but all channels are relevant (e.g. array-mic).

**Comment by pzelasko** [comment]
I agree with Dan, I'd rather keep Recording and Features fields separately.

Regarding array cuts: if I understood you correctly, we can do it in at least two ways right now:
- extract cuts separately for each channel and create a "MultiChannelASRDataset" (or however we name it) that sorts the CutSet by (recording-id, start-time, end-time, channel) tuple and retrieves all cuts for a given time span;
- create a mixed cut out of single-channel cuts and in the Dataset iterate the `mixed_cut.tracks` (which is somewhat hacky)

We could extend the support though, e.g.:
- add `load_tracks_features()` in `MixedCut` that returns a 3d array (n_channels, n_frames, n_features)... (and `load_tracks_audio()` probably)
- add utility that directly creates `MixedCut`s from `RecordingSet` with multiple `AudioSource`s

**Comment by YiwenShaoStephen** [comment]
Hi, I just found a potential issue in my recent experiments with raw waveform. It might not really about this PR but quite relevant so I just post it here. Please correct me if I misunderstand something.
As I notice, we are using torchaudio to extract features from raw waveform (according to here:https://github.com/pzelasko/lhotse/blob/master/lhotse/features.py#L122). 
But here are two potential issues with torchaudio, especially with torchaudio.compliance.kaldi:
- From my experiment, I found that these founctions can only work with CPU tensors (waveform). And in normal situations, we are most likely to put our data and model on GPUs and in that case this function will raise an error. (Haven't found any official document for this point and it would be great if someone can help me double check it).
- Waveform should be in shape (c, t) where c can only be in [0, 2) (https://pytorch.org/audio/compliance.kaldi.html#mfcc). In other words, it doesn't support batch level feature extraction and can only be used before you form a batch (e.g. in ```__getitem__```function). It would probably be OK in normal cases, but will make trouble if you'd like to do feature extraction outside dataset but in model forward instead.
As an alternative, ```torchaudio.transforms(torch.nn.Module)``` seems more flexible and we can even customize it as a normal ```nn.Module``` to fit with our user cases.
Please let me know how you think.
Thanks!

**Comment by danpovey** [comment]
Hi, I just found a potential issue in my recent experiments with raw
> waveform. It might not really about this PR but quite relevant so I just
> post it here. Please correct me if I misunderstand something.
> As I notice, we are using torchaudio to extract features from raw waveform
> (according to here:
> https://github.com/pzelasko/lhotse/blob/master/lhotse/features.py#L122).
> But here are two potential issues with torchaudio, especially with
> torchaudio.compliance.kaldi:
>
>    - From my experiment, I found that these founctions can only work with
>    CPU tensors (waveform). And in normal situations, we are most likely to put
>    our data and model on GPUs and in that case this function will raise an
>    error. (Haven't found any official document for this point and it would be
>    great if someone can help me double check it).
>    - Waveform should be in shape (c, t) where c can only be in [0, 2) (
>    https://pytorch.org/audio/compliance.kaldi.html#mfcc). In other words,
>    it doesn't support batch level feature extraction and can only be used
>    before you form a batch (e.g. in __getitem__function). It would
>    probably be OK in normal cases, but will make trouble if you'd like to do
>    feature extraction outside dataset but in model forward instead.
>    As an alternative, torchaudio.transforms(torch.nn.Module) seems more
>    flexible and we can even customize it as a normal nn.Module to fit
>    with our user cases.
>    Please let me know how you think.
>    Thanks!
>
>

Mm.
I wonder if it might make more sense to have the feature extraction as
python+PyTorch (or numpy?) code in lhotse itself?
The kaldi10feat project (on my github) may be something to refer to,
although it supports very little configuration.  It can't be made
compatible with current Kaldi features though, if that's a design criterion.
Dan

**Comment by danpovey** [comment]
... conventionally, people do compute the features on CPU, and for the most
common use case we'll probably extract features and write them (compressed)
to disk.  But of course it would be nice to be able to do it on GPU, in
general.

**Comment by pzelasko** [comment]
Thanks Yiwen. I'll open another issue for this.
